src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs

```diff
@@ -1223,6 +1231,76 @@ private async Task<SearchResult> SearchForReindexSurrogateIdsWithoutSearchParamH
             return searchResult;
         }
 
+        /// <summary>
+        /// Gets the count of resources within a specific surrogate ID range.
+        /// </summary>
+        /// <param name="resourceTypeId">The resource type ID</param>
+        /// <param name="startId">The lower bound surrogate ID (inclusive)</param>
+        /// <param name="endId">The upper bound surrogate ID (inclusive)</param>
+        /// <param name="searchParamHash">Optional search parameter hash filter</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>The count of resources within the specified range</returns>
+        private async Task<int> GetResourceCountBySurrogateIdRangeAsync(
+            short resourceTypeId,
+            long startId,
+            long endId,
+            string searchParamHash,
+            CancellationToken cancellationToken)
+        {
+            using var sqlCommand = new SqlCommand();
+            sqlCommand.CommandTimeout = GetReindexCommandTimeout();
+
+            if (!string.IsNullOrWhiteSpace(searchParamHash))
+            {
+                sqlCommand.Parameters.AddWithValue("@ResourceTypeId", resourceTypeId);
+                sqlCommand.Parameters.AddWithValue("@StartId", startId);
+                sqlCommand.Parameters.AddWithValue("@EndId", endId);
+                sqlCommand.Parameters.AddWithValue("@SearchParamHash", searchParamHash);
+
+                sqlCommand.CommandText = @"
+            SELECT COUNT(*) 
+            FROM dbo.Resource 
+            WHERE ResourceTypeId = @ResourceTypeId 
+              AND ResourceSurrogateId >= @StartId 
+              AND ResourceSurrogateId <= @EndId
+              AND IsHistory = 0 
+              AND IsDeleted = 0
+              AND (SearchParamHash != @SearchParamHash OR SearchParamHash IS NULL)";
+            }
+            else
+            {
+                sqlCommand.Parameters.AddWithValue("@ResourceTypeId", resourceTypeId);
+                sqlCommand.Parameters.AddWithValue("@StartId", startId);
+                sqlCommand.Parameters.AddWithValue("@EndId", endId);
+
+                sqlCommand.CommandText = @"
+            SELECT COUNT(*) 
+            FROM dbo.Resource 
+            WHERE ResourceTypeId = @ResourceTypeId 
+              AND ResourceSurrogateId >= @StartId 
+              AND ResourceSurrogateId <= @EndId
+              AND IsHistory = 0 
+              AND IsDeleted = 0";
+            }
+
+            LogSqlCommand(sqlCommand);
+
+            int count = 0;
+            await _sqlRetryService.ExecuteSql(
+                sqlCommand,
+                async (cmd, cancel) =>
+                {
+                    var result = await cmd.ExecuteScalarAsync(cancel);
+                    count = Convert.ToInt32(result);
```

Comment:
I'm concerned about using a 32bit int here. While a customer having billions of one resource type seems unlikely, is there any reason not to use a long here?